crates/sqllib/src/string_interner.rs

```diff
@@ -0,0 +1,616 @@
+    };
+
+    id
+}
+
+/// Returns the original string given its interned id.
+pub fn unintern_string(id: &InternedStringId) -> Option<SqlString> {
+    // Lookup the string in the cache first.
+    // If the string is not in the cache, look it up in the spine.
+    INTERNED_STRING_CACHE
+        .get(id)
+        .map(|(string, _step)| string)
+        .or_else(|| {
+            let mut cursor = INTERNED_STRING_BY_ID.read().unwrap().cursor();
+            if cursor.seek_key_exact(id) {
+                let val = unsafe { cursor.val().downcast::<Tup1<SqlString>>().0.clone() };
+                // Insert the string into the cache with pinned flag set to false, so it can be evicted.
+                INTERNED_STRING_CACHE.insert(id.clone(), (val.clone(), false));
+                Some(val)
+            } else {
+                None
+            }
+        })
+}
+
+/// Create an empty spine snapshot.
+fn empty_by_id() -> SpineSnapshot<DynOrdIndexedWSet<DynData, DynData, DynZWeight>> {
+    let factories: OrdIndexedWSetFactories<DynData, DynData, DynZWeight> =
+        BatchReaderFactories::new::<InternedStringId, Tup1<SqlString>, ZWeight>();
+
+    SpineSnapshot::<DynOrdIndexedWSet<DynData, DynData, DynZWeight>>::new(factories)
+}
+
+/// Build the string interner circuit.
+///
+/// Takes a stream that contains strings to be interned, and sets up a spine snapshot
+/// that maps interned string IDs back to strings. The spine snapshot is stored in the
+/// `INTERNED_STRING_BY_ID` global variable, and is updated on each step of the circuit.
+///
+// ```text
+//                                               ┌──────────┐                      ┌────────────────────────┐
+//                                               │integrate │                      │ Global interner state  │
+//                                               │  ┌───┐   │       ┌───────┐by_id │                        │
+//                                               │  │Z-1├───┼──────►│gather ├─────►│ INTERNED_STRINGS_BY_ID │
+//                                               │  └──┬┘   │       └───────┘      │ CURRENT_STEP           │
+//                                               │   ▲ │    │                      │ PINNED_STRINGS         │
+//                                               │   │ ▼    │                      └────────────────────────┘
+//   feldera_interned_strings ┌─────────┐        │  ┌┴──┐   │
+// ──────────────────────────►│map_index├────────┼─►│ + │   │
+//                            └─────────┘        │  └───┘   │
+//                                               └──────────┘
+// ```
+pub fn build_string_interner(interned_strings: Stream<RootCircuit, OrdZSet<Tup1<SqlString>>>) {
+    // Intern input strings, index them by interned string ID, and store them in a spine.
+    //
+    // The last step below `.delay_trace()` makes sure that we work with the spine snapshot
```

Comment:
how does this work with nested timestamps?
I assume the step is the outer circuit step
It would be nice to have an argument about why this is correct.

Reply:
It doesn't. It is currently the user's responsibility to collect all interned strings used by the program into a top-level stream. Strings that occur in a nested circuit must be exported to the top level.